src/mapWithFeedback.test.ts

```diff
@@ -0,0 +1,209 @@
+import { pipe } from "./pipe";
+import { mapWithFeedback } from "./mapWithFeedback";
+import { createLazyInvocationCounter } from "../test/lazy_invocation_counter";
+import { take } from "./take";
+import { add } from "./add";
+
+describe("data first", () => {
+  describe("base", () => {
+    it("should return an array of successively accumulated values", () => {
+      const result = mapWithFeedback(
+        [1, 2, 3, 4, 5] as const,
+        (acc, x) => acc + x,
+        100,
+      );
+      expect(result).toEqual([101, 103, 106, 110, 115]);
+    });
+
+    it("should use the same accumulator on every iteration if it's mutable, therefore returning an array containing {array length} references to the accumulator.", () => {
+      const results = mapWithFeedback(
+        [1, 2, 3, 4, 5] as const,
+        (acc, x) => {
+          acc[x] = x;
+          return acc;
+        },
+        {} as Record<string, unknown>,
+      );
+
+      const expectedEquality = Array.from({ length: 5 }).fill({
+        "1": 1,
+        "2": 2,
+        "3": 3,
+        "4": 4,
+        "5": 5,
+      });
+      expect(results).toEqual(expectedEquality);
+      const [item] = results;
+      expect(results.every((result) => result === item)).toBe(true);
+    });
+
+    it("if an empty array is provided, it should never iterate, returning an empty array.", () => {
+      const result = mapWithFeedback([], (acc) => acc, "value");
+      expect(result).toEqual([]);
```

Comment:
Also add a test against identity so nobody breaks this assumption
```suggestion
      const data = [];
      const result = mapWithFeedback(data, (acc) => acc, "value");
      expect(result).toEqual(data);
      expect(result).not.toBe(data);
```